### ISSUE TITLE
[Solved] Alaska2 branch: Mismatch lengths of signal array vs state vector with 10s epoch in spite of running the truncation routine

### ISSUE THREAD

**Issue by otoien** [paulbrodersen/somnotate#7]
The problems seems to be rooted in that my EDF data have a record length covering 5 seconds, thus sometimes having an odd number of records, while alignment at with _time_resolution = 10_  would require an even number of records in the data file.  This  is appearing with the sequence: _preprocess_ -  _truncate_ - _test state annotation_. Ignoring the error message from test state annotation and running training will give the same error and training stalls. However this is not consequent (odd and even referring to number of records in data file):
a) file1 even + file2 odd:  OK
b) file1 odd + file2 even:  Error
c) file1 even + file2 odd + file3 even:  OK
d) file1 even + file2 odd + file3 odd:  Error
At _time resolution = 5_ errors do not occur (expected).
At _time resolution = 10_ and only running the sequence  _preprocess_ - automated_state_annotation , errors do not occur.

Would it be possible to check for records that do not align to the set time resolution on load and truncate them in memory? I notice some code in _data_io.py_  under _def _load_edf_channels(signal_labels, edf_reader)_ :
```
total_samples = [edf_reader.samples_in_file(idx) for idx in indices]
    assert len(set(total_samples)) == 1, "All signals need to have the same length! Lengths of selected signals: {}".format(total_samples)
    total_samples, = set(total_samples)

    output_array = np.zeros((len(signal_labels), total_samples), dtype=np.int32)
    for jj, idx in enumerate(indices):
        edf_reader.read_digital_signal(idx, 0, total_samples, output_array[jj])
```
However this code appears to only check for different channels having the same length in the record?

The motivation for trying to make this work with _time resolution = 10_  is that current testing indicates, opposite of what was expected, that results have more accuracy relative to the manual scores at 10 sec than when tested with 5s or 1 sec time resolution.

I regret not truncating all files at the 10 sec border instead of at the 5 sec record when converting them to EDF. An alternative solution would be for me to write a dedicated routine to truncate  only the un-aligned test samples that we are manually scoring.

While this has been tested on my Alaska2 branch, the same behavior would perhaps be expected in the main branch, or at least in the alaska branch I used as starting point. 

Edit: I just noticed issue: https://github.com/paulbrodersen/somnotate/issues/4 which seems to be related. I missed it when posting as it was closed.

Edit2:   Paul @paulbrodersen , It is not as simple as outlined by a) and b) as even with swapping the odd and even file order in test a) I still got the error.

**Comment by otoien** [comment]
I think I have solved this. I loaded time_resolution into the truncate routine, and after the line reading total_time_in_seconds from the edf annotation file, I used the expression:
`total_time_in_seconds = time_resolution * (total_time_in_seconds // time_resolution)` to truncate the total time to a whole number of epochs. The data appears to already be similarly truncated on load. I was able to run 02_test_state_annotation on all the 18 data files sampled at 500 Hz in the Bear_selection folder without issue (not activating the plot). I will do some more checking before pushing this, so far only tested with time_resolution = 10s.
